pkg/preflight/run.go

```diff
@@ -0,0 +1,341 @@
+		signalChan := make(chan os.Signal, 1)
+		signal.Notify(signalChan, os.Interrupt)
+		<-signalChan
+		fmt.Print(cursor.Show())
+		os.Exit(0)
+	}()
+
+	var preflightContent []byte
+	var err error
+	if strings.HasPrefix(arg, "secret/") {
+		// format secret/namespace-name/secret-name
+		pathParts := strings.Split(arg, "/")
+		if len(pathParts) != 3 {
+			return errors.Errorf("path %s must have 3 components", arg)
+		}
+
+		spec, err := specs.LoadFromSecret(pathParts[1], pathParts[2], "preflight-spec")
+		if err != nil {
+			return errors.Wrap(err, "failed to get spec from secret")
+		}
+
+		preflightContent = spec
+	} else if _, err = os.Stat(arg); err == nil {
+		b, err := ioutil.ReadFile(arg)
+		if err != nil {
+			return err
+		}
+
+		preflightContent = b
+	} else {
+		if !util.IsURL(arg) {
+			return fmt.Errorf("%s is not a URL and was not found (err %s)", arg, err)
+		}
+
+		req, err := http.NewRequest("GET", arg, nil)
+		if err != nil {
+			return err
+		}
+		req.Header.Set("User-Agent", "Replicated_Preflight/v1beta2")
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+
+		preflightContent = body
+	}
+
+	preflightContent, err = docrewrite.ConvertToV1Beta2(preflightContent)
+	if err != nil {
+		return errors.Wrap(err, "failed to convert to v1beta2")
+	}
+	multidocs := strings.Split(string(preflightContent), "\n---\n")
+
+	troubleshootclientsetscheme.AddToScheme(scheme.Scheme)
+	decode := scheme.Codecs.UniversalDeserializer().Decode
+
+	var collectResults []preflight.CollectResult
+	preflightSpecName := ""
+	finishedCh := make(chan bool, 1)
+	progressCh := make(chan interface{}, 0) // non-zero buffer will result in missed messages
+
+	if v.GetBool("interactive") {
+		s := spin.New()
+		go func() {
+			lastMsg := ""
+			for {
+				select {
+				case msg, ok := <-progressCh:
+					if !ok {
+						continue
+					}
+					switch msg := msg.(type) {
+					case error:
+						c := color.New(color.FgHiRed)
+						c.Println(fmt.Sprintf("%s\r * %v", cursor.ClearEntireLine(), msg))
+					case string:
+						if lastMsg == msg {
+							break
+						}
+						lastMsg = msg
+						c := color.New(color.FgCyan)
+						c.Println(fmt.Sprintf("%s\r * %s", cursor.ClearEntireLine(), msg))
+					}
+				case <-time.After(time.Millisecond * 100):
+					fmt.Printf("\r  \033[36mRunning Preflight checks\033[m %s ", s.Next())
+				case <-finishedCh:
+					fmt.Printf("\r%s\r", cursor.ClearEntireLine())
+					return
+				}
+			}
+		}()
+	} else {
+		// make sure we don't block any senders
+		go func() {
+			for {
+				select {
+				case msg, ok := <-progressCh:
+					if !ok {
+						return
+					}
+					fmt.Fprintf(os.Stderr, "%v\n", msg)
+				case <-finishedCh:
+					return
+				}
+			}
+		}()
+	}
+
+	defer func() {
+		close(finishedCh)
+		close(progressCh)
+	}()
+
+	uploadResultsTo := ""
+	for _, doc := range multidocs {
```

Comment:
Some variable like `preflightSpecName` and `uploadResultsTo` are filled in every iteration. Are we interested only in the last element, because the values have been  overwrited in every iteration?

Reply:
This still needs fixing.  For our uses, we're only going to have a single spec and we won't rely on the multidocs code working.  Suggest we come back to it to fix or remove the multidoc code.

Reply:
My comment above is not strictly true - we do use multidocs, but the collection is working.  It's only the upload that may not be.